sapicore/engine/ensemble/__init__.py

```diff
@@ -0,0 +1,100 @@
+""" Ensembles (layers) manage neuron collections by fixing their size, extending select attribute tensors,
+and handling layer-aware operations such as heterogeneous parameter initialization and normalization.
+
+Note
+----
+Ensemble instance attributes are 1D tensors corresponding to potentially heterogeneous parameters
+and states of neurons (tensor elements) whose behavior is prescribed by their parent class
+(e.g., :class:`~engine.neuron.spiking.LIF.LIFNeuron`).
+
+Ensembles can be initialized programmatically or from a configuration YAML containing neuron parameters and their
+generation method (fixed, zipped, grid, or drawn from a distribution). Forward calls invoke the parent neuron class
+method.
+
+Example
+-------
+Initialize an ensemble of basic analog neurons, inheriting default parameters and behavior from
+:class:`~engine.neuron.analog.AnalogNeuron`:
+
+    >>> from sapicore.engine.ensemble.spiking import LIFEnsemble
+    >>> layer = LIFEnsemble(num_units = 10)
+
+See Also
+--------
+:class:`~utils.sweep.Sweep`
+    For built-in parameter sweep utility methods.
+
+"""
+import torch
+from torch import Tensor
+
+from sapicore.engine.neuron import Neuron
+
+
+class Ensemble(Neuron):
+    """Ensemble base class. Provides generic implementations of tensor expansion and parameter diversification.
+
+    Attributes
+    ----------
+    _extensible_props_: tuple[str]
+        The subset of this object's configurable property names (`_config_props_`) that needs to be duplicated
+        `num_units` times. Defaults to `None`, in which case all `_config_props_` are assumed extensible.
+        In the rare case where no configurable properties should be extended, use an empty tuple.
+
+    """
+
+    _extensible_props_: tuple[str] = None
+
+    def __init__(self, num_units: int = 1, **kwargs):
+        """Constructs a generic ensemble instance, inheriting attributes from :class:`~engine.neuron.Neuron`."""
+        super().__init__(**kwargs)
+        self._num_units = num_units
+
+        # expand extensible parameter tensors to size `num_units`.
+        for prop in self.extensible_props:
+            temp = getattr(self, prop)
+            setattr(self, prop, torch.zeros(self.num_units, dtype=torch.float, device=self.device) + temp)
+
+        # expand loggable property tensor buffers to size `num_units`.
+        for prop in self.loggable_props:
+            temp = getattr(self, prop)
+            setattr(self, prop, torch.zeros(self.num_units, dtype=torch.float, device=self.device) + temp)
+
+    @property
+    def num_units(self):
+        return self._num_units
+
+    @num_units.setter
+    def num_units(self, value: int):
+        self._num_units = value
+
+    @property
+    def extensible_props(self):
+        if self._extensible_props_ is None:
+            return self._config_props
+        else:
+            return self._extensible_props_
+
+    @extensible_props.setter
```

Comment:
I know you like having symmetry with things having both a setter and getter. But, for these kinds of properties I'd default to not having a setter if it's not intended to be used much.

Because, `_extensible_props_` and similarly `_config_props_` are meant to be listed by the class author and hardcode list the properties they wrote that should be configurable. That's not to say we disable changing it at runtime, but we shouldn't make easy with a setter so that end users don't try to update them dynamically. Because, this is mostly intended for our own internal usage and not for end users to mess about with.

Reply:
Removed setters